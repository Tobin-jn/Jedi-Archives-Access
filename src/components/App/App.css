.App {
  display: grid;
  grid-template-columns: 400px 1fr;
  height: 100vh;
}

aside {
  background-color: #000;
}

.button-section {
  display: flex;
  justify-content: center;
}

.category {
  text-align: center;
}