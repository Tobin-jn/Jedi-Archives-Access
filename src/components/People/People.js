import React from 'react'
import Card from '../Card/Card'
import './People.css'

const People = () => {
  return(
    <div className="people-container">

    </div>
  )
}

export default People