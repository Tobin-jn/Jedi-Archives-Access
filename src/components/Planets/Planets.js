import React from 'react'
import Card from '../Card/Card'
import './Planets.css'

const Planets = () => {
  return(
    <div className="p
    lanets-container">

    </div>
  )
}

export default Planets