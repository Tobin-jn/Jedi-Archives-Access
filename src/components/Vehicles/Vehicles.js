import React from 'react'
import Card from '../Card/Card'
import './Vehicles.css'


const Vehicles = () => {
  return(
    <div className="vehicles-container">

    </div>
  )
}

export default Vehicles