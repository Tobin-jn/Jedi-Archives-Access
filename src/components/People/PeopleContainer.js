import React from 'react'
import Card from '../Card/Card'
import './People.css'

const PeopleContainer = () => {
  return(
    <div className="people-container">

    </div>
  )
}

export default PeopleContainer