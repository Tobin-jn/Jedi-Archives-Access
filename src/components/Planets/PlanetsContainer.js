import React from 'react'
import Card from '../Card/Card'
import './Planets.css'

const PlanetsContainer = () => {
  return(
    <div className="planets-container">

    </div>
  )
}

export default PlanetsContainer