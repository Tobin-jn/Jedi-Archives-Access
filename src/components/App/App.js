import React, { Component } from 'react';
import './App.css';
import Crawl from '../Crawl/Crawl';
import Header from '../Header/Header';
import Button from '../Button/Button';
import CardContainer from '../CardContainer/CardContainer';
import fetchCall from '../../helpers/fetchCalls';
import Vehicles from '../../helpers/Vehicles';
import Planets from '../../helpers/Planets';
import People from '../../helpers/People';
import { Route, Switch, NavLink } from 'react-router-dom';
import Hamburger from '../Hamburger/Hamburger'


class App extends Component {
  constructor() {
    super()

    this.state = {
      currentSelection: '',
      openingCrawl: {},
      people: [],
      vehicles: [],
      planets: [],
      error: false,
      loading: true,
      fetchCall: fetchCall,
      fetchVehicles: new Vehicles(),
      fetchPeople: new People(),
      fetchPlanets: new Planets(),
      favorites: JSON.parse(localStorage.getItem('favorites')) || [],
      hamburger: 'closed',
      buttons: 'hide-buttons'
    };
  }

  componentDidMount() {
    this.crawlCall();
    this.checkStorage();
  }

  checkStorage = () => {
    this.setState({
        favorites: JSON.parse(localStorage.getItem('favorites')) || [],
        planets: JSON.parse(localStorage.getItem('planets')) || [],
        vehicles: JSON.parse(localStorage.getItem('vehicles')) || [],
        people: JSON.parse(localStorage.getItem('people')) || []
    })
  }

  hamburgerChange = () => {
    if (this.state.hamburger === 'closed') {
      this.setState({
        hamburger: 'deployed',
        buttons: 'deploy-buttons'
      });
    } else {
      this.setState({
        hamburger: 'closed',
        buttons: 'hide-buttons'
      });
    }
  }

  crawlCall = () => {
    if (localStorage.films) {
      this.getFilms();
    } else {
      this.fetchFilms();
    }
  }

  getFilms = () => {
    const films = JSON.parse(localStorage.getItem('films'));
    const randomNum = Math.floor(Math.random() * (films.count));
    this.setState({
      openingCrawl: films.results[randomNum], 
      loading: false,
      error: false
    })
  }

  fetchFilms = async () => {
    const url = 'https://swapi.co/api/films/';
    try {
      const films = await this.state.fetchCall(url);

      const randomNum = Math.floor(Math.random() * (films.count))
      this.setState({
        openingCrawl: films.results[randomNum], 
        loading: false,
        error: false
      })
      localStorage.setItem('films', JSON.stringify(films))
    } catch(error) {
      this.setState({ 
        error: true, 
        currentSelection: '' 
      })
    }
  }

  toggleFavorite = (cardData) => {
    this.toggleFavoriteInDatabase(cardData);
      
    if(this.state.favorites.find( fav => cardData.name === fav.name)){
      this.removeFavorite(cardData)


    } else {
      cardData.favorite = true;
      // console.log(cardData.favorite)
      const newFavorites = [...this.state.favorites, cardData]
      localStorage.setItem('favorites', JSON.stringify(newFavorites));
      this.setState({
        favorites: newFavorites,
        planets: JSON.parse(localStorage.getItem('planets')) || [],
        vehicles: JSON.parse(localStorage.getItem('vehicles')) || [],
        people: JSON.parse(localStorage.getItem('people')) || []
      })
    }
  }

  toggleFavoriteInDatabase = (cardData) => {
    const library = JSON.parse(localStorage.getItem(cardData.category))
    const target = library.find( card => cardData.name === card.name);
    target.favorite = !target.favorite;
    localStorage.setItem(cardData.category, JSON.stringify(library));
  }

  removeFavorite = (cardData) => {
    const updatedFavorites = this.state.favorites.filter( fav => fav.name !== cardData.name)
    localStorage.setItem('favorites', JSON.stringify(updatedFavorites))
    this.setState({
      favorites: updatedFavorites,
      planets: JSON.parse(localStorage.getItem('planets')) || [],
      vehicles: JSON.parse(localStorage.getItem('vehicles')) || [],
      people: JSON.parse(localStorage.getItem('people')) || []
    })
  }
 
  handleSelection = (currentSelection) => {
    if (currentSelection === 'people') {
      this.callFetchPeople();
    } else if (currentSelection === 'vehicles') {
      this.callFetchVehicles();
    } else if (currentSelection === 'planets') {
      this.callFetchPlanets()
    } else {
      this.setState({
        currentSelection: 'favorites',
        loading: false,
        error: false
      })
    }
  }

  callFetchVehicles = () => {
    if(localStorage.vehicles){
      this.pullVehicleData();
    } else {
      this.fetchVehicleData();
    }
  }

  pullVehicleData = () => {
    const response = localStorage.getItem('vehicles')
    const vehicles = JSON.parse(response)
    this.setState({
      vehicles: vehicles,
      currentSelection: 'vehicles',
      loading: false,
      error: false
    })    
  }

  fetchVehicleData = async() => {
    await this.setState({ loading: true })

    try {
      const cleanData = await this.state.fetchVehicles.fetchVehicles()
      this.setState({
        vehicles: cleanData,
        currentSelection: 'vehicles',
        loading: false,
        error: false
      })
      localStorage.setItem('vehicles', JSON.stringify(cleanData))
    } catch(error) {
      this.setState({ error: true, currentSelection: '' })
    }
  }

  callFetchPeople = () => {
    if (localStorage.people) {
      this.pullPeopleData();
    } else {
      this.fetchPeopleData();
    }
  }

  pullPeopleData = () => {
    const response = localStorage.getItem('people')
      const people = JSON.parse(response)
      this.setState({
        people: people,
        currentSelection: 'people',
        loading: false,
        error: false
    })
  }

  fetchPeopleData = async () => {
    await this.setState({ loading: true })

    try {
      const cleanedPeople = await this.state.fetchPeople.fetchPeople();
      this.setState({
        people: cleanedPeople,
        currentSelection: 'people',
        loading: false,
        error: false
      })
      localStorage.setItem('people', JSON.stringify(cleanedPeople))
    } catch(error) {
      this.setState({ error: true, currentSelection: '' })
    }
  }

  callFetchPlanets = () => {
    if (localStorage.planets){
      this.pullPlanetData()
    } else {
      this.fetchPlanetData()
    }
  }

  fetchPlanetData = async () => {
    await this.setState({ loading: true })
    try{
      const cleanedPlanets = await this.state.fetchPlanets.fetchPlanets()
      this.setState({
        planets: cleanedPlanets,
        currentSelection: 'planets',
        loading: false,
        error: false
      })
      localStorage.setItem('planets', JSON.stringify(cleanedPlanets))
    } catch(error) {
      this.setState({ error: true })
    }
  }

  pullPlanetData = () => {
    const response = localStorage.getItem('planets')
    const planets = JSON.parse(response)
    this.setState({
      planets: planets,
      currentSelection: 'planets',
      loading: false,
      error: false
    })
  }

  render() {
    return (
      <div className='App'>
        <div className="button-section">
          <header>
            <Hamburger hamburgerChange={this.hamburgerChange}/>
            <section className={`button-wrapper ${this.state.buttons}`}>
              <NavLink 
                to='/people' className='nav-button people' onClick={() => { this.handleSelection('people')}}>People</NavLink>
              <NavLink to='/planets' className='nav-button planets' onClick={() => { this.handleSelection('planets')}}>Planets</NavLink>
              <NavLink to='/vehicles' className='nav-button vehicles' onClick={() => { this.handleSelection('vehicles')}}>Vehicles</NavLink>
              <NavLink to='/favorites' className='nav-button favorite' onClick={() => { this.handleSelection('favorites')}}>Favorites</NavLink>
            </section>
          </header>
        </div>

        <Switch>
          <Route exact path='/' render={() => (
            <Crawl film={this.state.openingCrawl} />
          )} />

          <Route exact path='/people' render={() => (
            <CardContainer 
              data={this.state.people} 
              selection={this.state.currentSelection}
              toggleFavorite={this.toggleFavorite}
              error={this.state.error}
              loading={this.state.loading}
            />
          )} />

          <Route exact path='/planets' render={() => (
            <CardContainer 
              data={this.state.planets} 
              selection={this.state.currentSelection}
              toggleFavorite={this.toggleFavorite}
              error={this.state.error}
              loading={this.state.loading}
            />
          )} />

          <Route exact path='/vehicles' render={() => (
            <CardContainer 
              data={this.state.vehicles} 
              selection={this.state.currentSelection}
              toggleFavorite={this.toggleFavorite}
              error={this.state.error}
              loading={this.state.loading}
            />
          )} />

          <Route exact path='/favorites' render={() => (
            <CardContainer 
              data={this.state.favorites} 
              selection={this.state.currentSelection}
              toggleFavorite={this.toggleFavorite}
              error={this.state.error}
              loading={this.state.loading} 
            />
          )} />
        
        </Switch>
      </div>
    )
          // <Route component={404} />
  }
}

export default App;


