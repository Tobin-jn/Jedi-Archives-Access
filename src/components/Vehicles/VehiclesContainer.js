import React from 'react'
import Card from '../Card/Card'
import './Vehicles.css'

const VehiclesContainer = () => {
  return(
    <div className="vehicles-container">

    </div>
  )
}

export default VehiclesContainer