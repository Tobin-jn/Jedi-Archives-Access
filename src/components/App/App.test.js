import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import App from './App';
import fetchCall from '../../helpers/fetchCalls'
import Vehicles from '../../helpers/Vehicles'
import Planets from '../../helpers/Planets'
import People from '../../helpers/People'

describe('App', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />)
    wrapper.instance().fetchCall = jest.fn()
  });

  const defaultState = {
    totalFavorites: 0,
    currentSelection: '',
    openingCrawl: {},
    people: [],
    vehicles: [],
    planets: [],
    error: false,
    loading: true,
    fetchCall: fetchCall,
    fetchVehicles: new Vehicles(),
    fetchPeople: new People(),
    fetchPlanets: new Planets()
  };

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state', () => {
    expect(JSON.stringify(wrapper.state())).toEqual(JSON.stringify(defaultState));
  });

  it('should call crawlCall() on componentDidMount', () => {
    wrapper.instance().crawlCall = jest.fn()
    wrapper.instance().componentDidMount()
    expect(wrapper.instance().crawlCall).toHaveBeenCalled()
  });

  it('should call fetch when crawlCall is called', async () => {
    const mockFetch = jest.fn()
    wrapper.state().fetchCall = mockFetch
    await wrapper.instance().crawlCall()
    expect(mockFetch).toHaveBeenCalled()

  });

  it('should set state when crawlCall is called', async () => {
    const mockFetch = jest.fn(() => { return {
      results: [{film: 'test'}, {film: 'test'}], count: 2
    }});
    wrapper.state().fetchCall = mockFetch
    await wrapper.instance().crawlCall()

    expect(wrapper.state().openingCrawl).toEqual({film: 'test'})
  });

  it('should call callFetchPeople if people is currentSelection', () => {
    const mockCurrentSelection = 'people';
    wrapper.instance().callFetchPeople = jest.fn();

    wrapper.instance().handleSelection(mockCurrentSelection)
    expect(wrapper.instance().callFetchPeople).toHaveBeenCalled();
  });

  it('should call callFetchVehicles if vehicles is currentSelection', () => {
    const mockCurrentSelection = 'vehicles';
    wrapper.instance().callFetchVehicles = jest.fn();

    wrapper.instance().handleSelection(mockCurrentSelection)
    expect(wrapper.instance().callFetchVehicles).toHaveBeenCalled();
  });

  it('should call callFetchPlanets if planets is currentSelection', () => {
    const mockCurrentSelection = 'planets';
    wrapper.instance().callFetchPlanets = jest.fn();

    wrapper.instance().handleSelection(mockCurrentSelection)
    expect(wrapper.instance().callFetchPlanets).toHaveBeenCalled();
  });

  it('should call fetchVehicles if callFetchVehicles is called', async () => {
    const mockFetch = jest.fn();
    const mockVehicleClass = {fetchVehicles: mockFetch}
    wrapper.state().fetchVehicles = mockVehicleClass;

    await wrapper.instance().callFetchVehicles();
    expect(mockFetch).toHaveBeenCalled();
  });

  it('should set state after callFetchVehicles is called', async () => {
    const mockFetch = jest.fn(() => {
      return [{vehicle: 'car'}, {vehicle: 'boat'}]
    });
    const mockVehicleClass = {fetchVehicles: mockFetch}
    wrapper.state().fetchVehicles = mockVehicleClass;
    
    await wrapper.instance().callFetchVehicles();

    expect(wrapper.state().vehicles).toEqual([{vehicle: 'car'}, {vehicle: 'boat'}])
  });

  it('should call fetchPeople if callFetchPeople is called', async () => {
    const mockFetch = jest.fn();
    const mockPeopleClass = {fetchPeople: mockFetch}
    wrapper.state().fetchPeople = mockPeopleClass;

    await wrapper.instance().callFetchPeople();
    expect(mockFetch).toHaveBeenCalled();
  });

  it('should set state after callFetchPeople is called', async () => {
    const mockFetch = jest.fn(() => {
      return [{person: 'joe'}, {person: 'sara'}]
    });
    const mockPeopleClass = {fetchPeople: mockFetch}
    wrapper.state().fetchPeople = mockPeopleClass;
    
    await wrapper.instance().callFetchPeople();

    expect(wrapper.state().people).toEqual([{person: 'joe'}, {person: 'sara'}])
  });

  it('should call fetchPlanets if callFetchPlanets is called', async () => {
    const mockFetch = jest.fn();
    const mockPlanetsClass = {fetchPlanets: mockFetch}
    wrapper.state().fetchPlanets = mockPlanetsClass;

    await wrapper.instance().callFetchPlanets();
    expect(mockFetch).toHaveBeenCalled();
  });

  it('should set state after callFetchPlanets is called', async () => {
    const mockFetch = jest.fn(() => {
      return [{planet: 'earth'}, {planet: 'mars'}]
    });
    const mockPlanetClass = {fetchPlanets: mockFetch}
    wrapper.state().fetchPlanets = mockPlanetClass;
    
    await wrapper.instance().callFetchPlanets();

    expect(wrapper.state().planets).toEqual([{planet: 'earth'}, {planet: 'mars'}])
  });

  it('should set an error on state after a failed crawlCall call', async () => {
    const mockFetch = jest.fn().mockImplementation(() =>
      Promise.reject());
    wrapper.state().fetchCall = mockFetch;

    await wrapper.instance().crawlCall();

    expect(wrapper.state().error).toEqual(true);
  });

  it('should set an error on state after a failed fetchPeople call', async () => {
    const mockFetch = jest.fn().mockImplementation(() =>
      Promise.reject());
    wrapper.state().fetchPeople = mockFetch;

    await wrapper.instance().callFetchPeople();

    expect(wrapper.state().error).toEqual(true);
  });

  it('should set an error on state after a failed fetchVehicles call', async () => {
    const mockFetch = jest.fn().mockImplementation(() =>
      Promise.reject());
    wrapper.state().fetchVehicles = mockFetch;

    await wrapper.instance().callFetchVehicles();

    expect(wrapper.state().error).toEqual(true);
  });

  it('should set an error on state after a failed fetchPlanets call', async () => {
    const mockFetch = jest.fn().mockImplementation(() =>
      Promise.reject());
    wrapper.state().fetchPlanets = mockFetch;

    await wrapper.instance().callFetchPlanets();

    expect(wrapper.state().error).toEqual(true);
  });
})
