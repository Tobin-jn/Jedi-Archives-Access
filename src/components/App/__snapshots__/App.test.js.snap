// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="App"
>
  <div
    aria-label="navigation-section-wrapper"
    className="button-section"
  >
    <header>
      <Hamburger
        hamburgerChange={[Function]}
        status={
          Object {
            "bottomBladeToggled": "",
            "bottomToggled": "",
            "gripToggled": "",
            "middleBladeToggled": "",
            "middleToggled": "",
            "status": "closed",
            "topBladeToggled": "",
            "topToggled": "",
          }
        }
      />
      <section
        aria-label="mobile-navigation-section"
        className="modal-wrapper"
      >
        <ModalButtons
          currentSelection=""
          hamburger={
            Object {
              "bottomBladeToggled": "",
              "bottomToggled": "",
              "gripToggled": "",
              "middleBladeToggled": "",
              "middleToggled": "",
              "status": "closed",
              "topBladeToggled": "",
              "topToggled": "",
            }
          }
          handleSelection={[Function]}
        />
      </section>
      <h4
        className="please-login "
      >
        Please login to access the archives
      </h4>
      <section
        aria-label="navigation"
        className="button-wrapper hide-buttons"
      >
        <NavLink
          activeClassName="active"
          aria-current="page"
          className="nav-button people"
          to="/people"
        >
          <Button
            buttonName="people"
            currentSelection=""
            handleSelection={[Function]}
          />
        </NavLink>
        <NavLink
          activeClassName="active"
          aria-current="page"
          className="nav-button planets"
          to="/planets"
        >
          <Button
            buttonName="planets"
            currentSelection=""
            handleSelection={[Function]}
          />
        </NavLink>
        <NavLink
          activeClassName="active"
          aria-current="page"
          className="nav-button vehicles"
          to="/vehicles"
        >
          <Button
            buttonName="vehicles"
            currentSelection=""
            handleSelection={[Function]}
          />
        </NavLink>
        <NavLink
          activeClassName="active"
          aria-current="page"
          className="nav-button favorites"
          to="/favorites"
        >
          <Button
            buttonName="saved"
            currentSelection=""
            handleSelection={[Function]}
          />
        </NavLink>
      </section>
    </header>
  </div>
  <Switch>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/people"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/planets"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/vehicles"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/favorites"
      render={[Function]}
    />
  </Switch>
</div>
`;
